### PR TITLE
[OV2.0] Preprocessing - resize

### PR DESCRIPTION
**Pull request by nosovmik**
### Details:
 - TensorInfo API: set_spacial_dynamic_shape, set_spacial_static_shape(height, width)
 - PreprocessSteps::resize() - default (to network dimension)
 - PreprocessSteps::resize(height, width) - version with specified dimensions
 - Network().set_layout() API is introduced to set layout of network's tensor. Needed to determine width&height for scaling
 - Reference tests with hardcoded values are implemented on template plugin
 - CPU tests: added tests to compare with reference template plugin implementation
 - nGraph unit tests - error cases for resize
 - Test coverage is 100%

### Leftovers (dependency on other API): 
 - Several test cases will need to be implemented once convert_color and convert_layout is available

### Tickets:
 - 62185

### CODE REVIEW

File: ngraph/core/src/preprocess/preprocess_steps_impl.cpp [hunk 1]
```diff
@@ -97,3 +90,66 @@ void PreProcessSteps::PreProcessStepsImpl::add_convert_impl(const ov::element::T
...
+                            "of image data, like `NCHW`");
+            auto node_rank = node->get_output_partial_shape(0).rank();
+            OPENVINO_ASSERT(node_rank.is_static(), "Resize operation is not supported for fully dynamic shape");
+
+            auto height_idx = ov::layout::height(layout);
+            if (height_idx < 0) {  // E.g. height_idx = -1 means last dimension
+                height_idx = node_rank.get_length() + height_idx;
+            }
+            OPENVINO_ASSERT(height_idx >= 0 && height_idx < node_rank.get_length(),
+                            "Height dimension is out of bounds");
```
[the remark this answers is not in the document]
Comment (nosovmik) [reply ngraph/core/src/preprocess/preprocess_steps_impl.cpp]: Fixed

File: docs/template_plugin/tests/functional/op_reference/base_reference_test.hpp [hunk 2]
```diff
@@ -67,6 +85,11 @@ struct Tensor {
     Tensor(const ngraph::Shape& shape, ngraph::element::Type type, const std::vector<T>& data_elements)
         : Tensor {shape, type, CreateBlob(type, data_elements)} {}
 
+    // Use this constructor of dynamic network inputs
+    template <typename T>
+    Tensor(ngraph::element::Type type, const ngraph::Shape& shape, const std::vector<T>& data_elements)
```
Comment (ilyachur) [inline docs/template_plugin/tests/functional/op_reference/base_reference_test.hpp:90]: Why do we need to have new constructor with different order of arguments?

File: ngraph/core/src/preprocess/pre_post_process.cpp [hunk 3]
```diff
@@ -265,6 +384,26 @@ PreProcessSteps&& PreProcessSteps::convert_element_type(const element::Type& typ
     return std::move(*this);
 }
 
+PreProcessSteps& PreProcessSteps::resize(ResizeAlgorithm alg, size_t dst_height, size_t dst_width) & {
+    m_impl->add_resize_impl(alg, static_cast<int>(dst_height), static_cast<int>(dst_width));
```
Comment (ilyachur) [inline ngraph/core/src/preprocess/pre_post_process.cpp:388]: What if `dst_height > max<int>`?

File: docs/template_plugin/tests/functional/op_reference/base_reference_test.hpp (hunk 2 above)
Comment (nosovmik) [reply docs/template_plugin/tests/functional/op_reference/base_reference_test.hpp:90]: In these tests blob is created based on input shape. However for 'resize' the common case is that width&height are dynamic. So, test blob should be created based on 'input data elements shape', not based on 'parameter shape'
I didn't find easier solution, so decided to introduce one more constructor  and one more variant of 'CreateBlob' with specified shape

Comment (ilya-lavrenov) [reply docs/template_plugin/tests/functional/op_reference/base_reference_test.hpp:90]: I also don't understand why do we need new ctor..

File: docs/template_plugin/tests/functional/op_reference/base_reference_test.cpp [hunk 4]
```diff
@@ -32,13 +32,40 @@ void CommonReferenceTest::LoadNetwork() {
...
+//        InferenceEngine::Blob::Ptr blob;
+//        if (!info->getTensorDesc().getDims().empty() && info->getTensorDesc().getDims()[0] != 0) {
+//            blob = make_blob_with_precision(info->getTensorDesc());
+//        } else {
+//            blob = make_plain_blob(info->getTensorDesc().getPrecision(), inputData[i]->getTensorDesc().getDims());
+//        }
+//        blob->allocate();
+//
+//        ASSERT_EQ(blob->byteSize(), inputData[i]->byteSize());
+//=======
```
Comment (ilya-lavrenov) [inline docs/template_plugin/tests/functional/op_reference/base_reference_test.cpp:56]: do we need the code above?

**Review by ilya-lavrenov — APPROVED** [review]

File: docs/template_plugin/tests/functional/subgraph_reference/preprocess.cpp [hunk 5]
```diff
@@ -250,6 +250,155 @@ static RefPreprocessParams mean_scale_dynamic_layout() {
...
+}
+
+static RefPreprocessParams resize_to_specified_width_height() {
+    RefPreprocessParams res("resize_to_specified_width_height");
+    res.function = []() {
+        auto f = create_simple_function(element::f32, PartialShape{1, 1, Dimension::dynamic(), Dimension::dynamic()});
+        f = PrePostProcessor()
+                .input(InputInfo()
+                               .tensor(InputTensorInfo().set_spatial_dynamic_shape())
+                               .preprocess(PreProcessSteps().resize(ResizeAlgorithm::RESIZE_NEAREST, 4, 4))
```
[the remark this answers is not in the document]
Comment (nosovmik) [reply docs/template_plugin/tests/functional/subgraph_reference/preprocess.cpp:366]: yes, it is called at the end of 'build'

File: docs/template_plugin/tests/functional/op_reference/base_reference_test.hpp (hunk 2 above)
Comment (nosovmik) [reply docs/template_plugin/tests/functional/op_reference/base_reference_test.hpp]: After rebase, still had to introduce one more constructor as modifying this causes a lot of unexpected failures of other tests

File: inference-engine/tests/ngraph_helpers/ngraph_functions/include/ngraph_functions/preprocess/preprocess_builders.hpp [hunk 6]
```diff
@@ -192,6 +192,42 @@ inline std::shared_ptr<Function> tensor_layout() {
...
+}
+
+inline std::shared_ptr<Function> resize_nearest() {
+    using namespace ov::preprocess;
+    auto function = create_preprocess_1input(element::f32, PartialShape{1, 3, 10, 10});
+    function = PrePostProcessor()
+            .input(InputInfo()
+                           .tensor(InputTensorInfo().set_spatial_static_shape(20, 20))
+                           .preprocess(PreProcessSteps().resize(ResizeAlgorithm::RESIZE_NEAREST))
+                           .network(InputNetworkInfo().set_layout("NCHW")))
```
[the remark this answers is not in the document]
Comment (nosovmik) [reply inference-engine/tests/ngraph_helpers/ngraph_functions/include/ngraph_functions/preprocess/preprocess_builders.hpp]: Done. Added NHWC case to next test. Also added 'NHWC' test to 'template plugin subgraph-reference tests'

File: docs/template_plugin/tests/functional/op_reference/base_reference_test.cpp (hunk 4 above)
Comment (nosovmik) [reply docs/template_plugin/tests/functional/op_reference/base_reference_test.cpp]: Removed, thanks

File: ngraph/core/src/preprocess/pre_post_process.cpp (hunk 3 above)
Comment (nosovmik) [reply ngraph/core/src/preprocess/pre_post_process.cpp:391]: Added assert for this

**Review by ilyachur — APPROVED** [review]

**Pull request merged**